File: alpha/experiments/item-storage-protocol-standalone/src/clock.ts
// A simulated clock. Everything time-stamped in the ledgers reads its "now" from
// here, and time only advances when an experiment explicitly advances it. That
// keeps timestamps reproducible (no Date.now()) and lets rent — measured in
// byte-days — be integrated over a known, deterministic timeline.

const MS_PER_DAY = 24 * 60 * 60 * 1000;

export class SimClock {
  /** Whole days elapsed since the fixed epoch. */
  private day: number;
  private readonly epochMs: number;

  constructor(epochISO = "2020-01-01T00:00:00.000Z") {
    this.epochMs = Date.parse(epochISO);
    this.day = 0;
  }

  now(): number {
    return this.day;
  }

  iso(): string {
    return new Date(this.epochMs + this.day * MS_PER_DAY).toISOString();
  }

  advanceDays(n: number): void {
    if (n < 0) throw new Error("time does not run backward in this simulation");
    this.day += n;
  }
}

File: alpha/experiments/item-store/src/clock.rs
//! A deterministic day clock. Time only advances when explicitly told to, so
//! timestamps and the byte-day rent integral are reproducible run to run (no
//! wall-clock reads).
//!
//! Ports `item-storage-protocol-standalone/src/clock.ts` (the day-counter core;
//! the ISO helper is omitted — the ledger treats timestamps as opaque strings).

/// A simulated clock measured in whole days since a fixed epoch.
#[derive(Debug, Default)]
pub struct SimClock {
    day: u64,
}

impl SimClock {
    /// A new clock at day 0.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// The current day.
    #[must_use]
    pub fn now(&self) -> u64 {
        self.day
    }

    /// Advance the clock by `n` days. Time cannot run backward (`n: u64`).
    pub fn advance_days(&mut self, n: u64) {
        self.day += n;
    }
}

#[cfg(test)]
mod tests {
    use super::SimClock;

    #[test]
    fn advances_monotonically_from_zero() {
        let mut clock = SimClock::new();
        assert_eq!(clock.now(), 0);
        clock.advance_days(30);
        assert_eq!(clock.now(), 30);
        clock.advance_days(5);
        assert_eq!(clock.now(), 35);
    }
}
